core/commonMain/src/kotlinx/serialization/encoding/ReorderingCompositeEncoder.kt

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright 2017-2024 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization.encoding
+
+import kotlinx.serialization.*
+import kotlinx.serialization.descriptors.*
+import kotlinx.serialization.modules.*
+
+
+/**
+ * Encodes composite elements in a specific order managed by [mapElementIndex].
+ *
+ * This encoder will replicate the behavior of a standard encoding, but calling the `encode*Element` methods in
+ * the order defined by [mapElementIndex]. It first buffers each `encode*Element` calls in an array following
+ * the given indexes using [mapElementIndex], then when [endStructure] is called, it encodes the buffered calls
+ * in the expected order by replaying the previous calls on a concrete [CompositeEncoder] provided
+ * by [beginStructure].
+ *
+ * @param structureDescriptor descriptor of the structure being encoded
+ * @param beginStructure provides the [CompositeEncoder] to be used to encode the given descriptor's elements in the expected order.
+ * @param mapElementIndex maps the element index to the new reordered index (zero-based). If this mapper provides the same index for multiple elements, only the last one will be encoded as the previous ones will be overridden.
+ */
+@ExperimentalSerializationApi
+internal class ReorderingCompositeEncoder(
+    structureDescriptor: SerialDescriptor,
+    private val beginStructure: (SerialDescriptor) -> CompositeEncoder,
+    private val mapElementIndex: (SerialDescriptor, Int) -> Int,
+) : CompositeEncoder {
+    // Each time we encode a field, if the next expected schema field index is not the good one, it is buffered until it's the time to encode it
+    private var bufferedCalls = Array<BufferedCall?>(structureDescriptor.elementsCount) { null }
```

Comment:
This cache means you need to create an instance at least per structure descriptor, but more conveniently for every instance. In such case `beginStructure` can just become `private val delegateEncoder: CompositeEncoder`.

Although it may also make sense to have this retained (and used for subStructures, although in such case the element index would be an important parameter)

Reply:
For the first part, I suppose you were talking about the supplier that should just be the encoder (no supplier), and I agree.

But for the second part, I don't understand, the buffer is just for the corresponding descriptor, that should not be shared with sub-structure: we are reordering one level of a given descriptor, but not the sub structures.